### ISSUE TITLE
MA0095 - A class that implements IEquatable<T> should override Equals(object) is handled by CA1067

### ISSUE THREAD

**Issue by Meir017** [meziantou/Meziantou.Analyzer#857]
- **Rule**: [MA0095](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0095.md) - A class that implements IEquatable<T> should override Equals(object)
- **Built-in Equivalent**: [CA1067](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1067) - Override Equals when implementing IEquatable

## Analysis

**MA0095** and **CA1067** serve the same purpose: ensuring that types implementing `IEquatable<T>` also override the base `Equals(object)` method to provide consistent equality behavior across both generic and non-generic contexts.

### MA0095 Coverage
- Detects IEquatable<T> implementations without Equals(object) override
- Ensures consistent equality semantics
- Promotes proper equality implementation patterns

### CA1067 Coverage  
- Detects types implementing IEquatable<T> that should override Equals
- Official Microsoft rule for proper equality implementation
- Ensures equality consistency across different contexts
- Part of the .NET design guidelines

## Code Examples

**Problematic code detected by both rules:**
```csharp
// Both MA0095 and CA1067 will flag this
public class Person : IEquatable<Person>
{
    public string Name { get; set; }
    public int Age { get; set; }
    
    // Implements IEquatable<Person>
    public bool Equals(Person other)
    {
        if (other == null) return false;
        return Name == other.Name && Age == other.Age;
    }
    
    // Missing! Should override Equals(object)
    // This causes inconsistent behavior!
    
    public override int GetHashCode()
    {
        return HashCode.Combine(Name, Age);
    }
}
```

**Recommended fix:**
```csharp
// Both rules suggest overriding Equals(object)
public class Person : IEquatable<Person>
{
    public string Name { get; set; }
    public int Age { get; set; }
    
    // Generic equality implementation
    public bool Equals(Person other)
    {
        if (other == null) return false;
        return Name == other.Name && Age == other.Age;
    }
    
    // Object equality implementation - now consistent!
    public override bool Equals(object obj)
    {
        return Equals(obj as Person);
    }
    
    public override int GetHashCode()
    {
        return HashCode.Combine(Name, Age);
    }
    
    // Optional but recommended
    public static bool operator ==(Person left, Person right)
    {
        return EqualityComparer<Person>.Default.Equals(left, right);
    }
    
    public static bool operator !=(Person left, Person right)
    {
        return !(left == right);
    }
}
```

## Inconsistency Problems

Without proper Equals(object) override:
```csharp
var person1 = new Person { Name = "John", Age = 30 };
var person2 = new Person { Name = "John", Age = 30 };

// These give different results - inconsistent!
bool generic = person1.Equals(person2);        // true (uses IEquatable<T>)
bool objectBased = person1.Equals((object)person2); // false (uses Object.Equals)

// Collections may behave unexpectedly
var set = new HashSet<object> { person1 };
bool contains = set.Contains(person2); // May be false due to Object.Equals!
```

## Benefits of Consistent Implementation

Proper equality implementation provides:
- **Consistent behavior** across generic and non-generic contexts
- **Reliable collection operations** in mixed-type scenarios
- **Predictable LINQ operations** when using object-based comparisons
- **Better polymorphic behavior** in inheritance hierarchies

## Recommendation

**CA1067** is the official Microsoft analyzer rule that provides **equivalent functionality** to MA0095. Since both rules target the same equality consistency issue and CA1067 is the standard rule for equality implementation, **MA0095 should be deprecated** in favor of CA1067.

## Migration Path

1. Enable **CA1067** in your analyzer configuration:
   ```xml
   <WarningsAsErrors Include="CA1067" />
   ```
2. Disable **MA0095**:
   ```xml
   <WarningsNotAsErrors Include="MA0095" />
   ```
3. Update any rule suppressions from MA0095 to CA1067

## Related

- Parent issue: #835 - Consider deprecating rules that are now covered by built-in .NET analyzers

**Comment by Meir017** [comment]
## Analysis: Equivalent Rules - Batch 1 ✅

**MA0095 vs CA1067** - Equals override consistency
**MA0102 vs IDE0044** - Readonly field optimization  
**MA0112 vs CA1860** - Collection performance optimization
**MA0131 vs CA2264** - ThrowIfNull validation
**MA0142 vs IDE0041** - Pattern matching modernization

### Recommendation: Deprecate All ✅

Each has equivalent Microsoft analyzer counterpart.

---
*Part of #835 deprecation initiative.*

**Comment by Meir017** [comment]
This issue should be closed - MA0069 equivalency is already documented in [docs/comparison-with-other-analyzers.md](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/comparison-with-other-analyzers.md) - MA0069 is equivalent to CA2211.
